### PR TITLE
 Add ability to configure OAuth type in Multi User Che

### PR DESCRIPTION
**Pull request by mkuznyetsov**
### What does this PR do?
Decoupled implementation of OAuthAuthentication service into a separate component, which allows it's binding depending on configuration property (`che.oauth.mode`).
There are two modes that can be configured:
- `embedded` (using Che's OAuthAuthenticator), which is default, and only mode for single user Che.
- `delegated` (using Keycloak IdentityProviders) previously used by KeycloakOAuthAuthenticationService. This mode is default to multi user mode.
### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/9461

### CODE REVIEW

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties [hunk 1]
```diff
@@ -136,3 +136,9 @@ che.keycloak.js_adapter_url=NULL
 #     a discovery endpoint as detailed in the following specification
 #     https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig
 che.keycloak.oidc_provider=NULL
+
+#     Configuration of OAuth Authentication Service that can be used in "embedded" or "delegated" mode.
+#     If set to "embedded", then the service work as a wrapper to Che's OAuthAuthenticator ( as in Single User mode).
+#     If set to "delegated", then the service will use Keycloak IdentityProvider mechanism.
+#     Delegated mode will be used by default, in case if this property is not set properly.
+che.oauth.service.mode=embedded
```
Comment (skabashnyuk) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: new line

**Comment by skabashnyuk** [comment]
ci-test

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/oauth2/DelegatedOAuthAPI.java [hunk 2]
```diff
@@ -13,54 +13,42 @@
...
 import org.eclipse.che.multiuser.keycloak.shared.dto.KeycloakTokenResponse;
+import org.eclipse.che.security.oauth.OAuthAPI;
+import org.eclipse.che.security.oauth.shared.dto.OAuthAuthenticatorDescriptor;
 
-@Path("/oauth")
-public class KeycloakOAuthAuthenticationService {
-  @Context UriInfo uriInfo;
-
-  @Context SecurityContext security;
+/** */
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/oauth2/DelegatedOAuthAPI.java]: Please fix java doc

File: wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth/OAuthAPI.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  Response authenticate(
+      UriInfo uriInfo,
+      String oauthProvider,
+      List<String> scopes,
+      String redirectAfterLogin,
+      HttpServletRequest request)
+      throws NotFoundException, OAuthAuthenticationException, ForbiddenException,
+          BadRequestException;
+
+  Response callback(UriInfo uriInfo, List<String> errorValues)
```
Comment (sleshchenko) [inline wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth/OAuthAPI.java]: Please add java doc

File: wsmaster/che-core-api-auth/src/test/java/org/eclipse/che/security/oauth/EmbeddedOAuthAPITest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.eclipse.che.api.auth.shared.dto.OAuthToken;
+import org.eclipse.che.api.core.NotFoundException;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.testng.MockitoTestNGListener;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+/** @author Max Shaposhnik */
```
Comment (sleshchenko) [inline wsmaster/che-core-api-auth/src/test/java/org/eclipse/che/security/oauth/EmbeddedOAuthAPITest.java]: Is Max really an author of this class?

**Review by vparfonov — APPROVED** [review]

**Comment by mkuznyetsov** [comment]
ci-test

**Comment by mkuznyetsov** [comment]
ci-test

**Comment by mkuznyetsov** [comment]
ci-test

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties [hunk 5]
```diff
@@ -136,3 +136,9 @@ che.keycloak.js_adapter_url=NULL
 #     a discovery endpoint as detailed in the following specification
 #     https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig
 che.keycloak.oidc_provider=NULL
+
+#     Configuration of OAuth Authentication Service that can be used in "embedded" or "delegated" mode.
+#     If set to "embedded", then the service work as a wrapper to Che's OAuthAuthenticator ( as in Single User mode).
+#     If set to "delegated", then the service will use Keycloak IdentityProvider mechanism.
+#     Delegated mode will be used by default, in case if this property is not set properly.
+che.oauth.service_mode=delegated
```
Comment (skabashnyuk) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: new line

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/OAuthAPIProvider.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ */
+public class OAuthAPIProvider implements Provider<OAuthAPI> {
+  private static final Logger LOG = LoggerFactory.getLogger(OAuthAPIProvider.class);
+  private String oauthType;
+  private Injector injector;
+
+  @Inject
+  public OAuthAPIProvider(
+      @Nullable @Named("che.oauth.service_mode") String oauthType, Injector injector) {
+    this.oauthType = oauthType == null ? "" : oauthType;
```
Comment (skabashnyuk) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/OAuthAPIProvider.java]: I guess we need to fail here if oauthType!= embedded or keycloak

**Pull request merged**